File: sem4/mpp/mpp-proiect-csharp-vreauladudu/repository/database/AbstractDatabase.cs
using mpp_proiect_csharp_vreauladudu.domain;

namespace mpp_proiect_csharp_vreauladudu.repository.database
{
    internal abstract class AbstractDatabase<ID, E> : IRepository<ID, E> where E : Entity<ID>
    {
        public String? host { get; set; }
        public String? username { get; set; }
        public String? password { get; set; }
        public String? table { get; set; }

        protected AbstractDatabase(string? host, string? username, string? password, string? table)
        {
            this.host = host;
            this.username = username;
            this.password = password;
            this.table = table;
        }

        public E? find(ID id)
        {
            if (id == null)
            {
                throw new ArgumentNullException("entity must not be null");
            }

            return null;
        }

        public IEnumerable<E> findAll()
        {
            return null;
        }

        public E? save(E e)
        {
            return null;
        }

        public E? remove(ID id)
        {
            if (id == null)
            {
                throw new ArgumentNullException("entity must not be null");
            }

            return null;
        }
        public E? update(E e)
        {
            return null;
        }
    }
}

File: sem4/mpp/mpp-proiect-java-vreauladudu/persistence/src/main/java/moto/repository/database/AbstractDatabase.java
package moto.repository.database;

import moto.Entity;
import moto.repository.IRepository;

import java.sql.*;
import java.util.HashSet;
import java.util.Properties;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

public abstract class AbstractDatabase<ID, E extends Entity<ID>> implements IRepository<ID, E>
{
    private final String table;
    private static final Logger logger = LogManager.getLogger();

    public AbstractDatabase(Properties properties, String table)
    {
        DatabaseUtils.setProperties(properties);
        this.table = table;
    }

    protected abstract E extractEntity(ResultSet resultSet) throws SQLException;

    protected abstract PreparedStatement findStatement(ID id) throws SQLException;
    protected abstract PreparedStatement saveStatement(E entity) throws SQLException;
    protected abstract PreparedStatement deleteStatement(ID id) throws SQLException;
    protected abstract PreparedStatement updateStatement(E entity) throws SQLException;

    @Override
    public E find(ID id) throws IllegalArgumentException
    {
        logger.traceEntry();
        if (id == null)
        {
            throw new IllegalArgumentException("id must not be null");
        }

        try
        {
            ResultSet resultSet = findStatement(id).executeQuery();
            if (resultSet.next())
            {
                E entity = this.extractEntity(resultSet);
                logger.traceExit(entity);
                return entity;
            }
        }
        catch (SQLException exception)
        {
            logger.error(exception);
            exception.printStackTrace();
        }
        logger.traceExit();
        return null;
    }

    @Override
    public Iterable<E> findAll()
    {
        logger.traceEntry();
        HashSet<E> entities = new HashSet<>();
        try
        {
            ResultSet resultSet = DatabaseUtils.getConnection().prepareStatement("SELECT * FROM " + this.table).executeQuery();
            while (resultSet.next())
            {
                entities.add(this.extractEntity(resultSet));
            }
        }
        catch (SQLException exception)
        {
            logger.error(exception);
            exception.printStackTrace();
        }
        logger.traceExit(entities);
        return entities;
    }

    @Override
    public E save(E entity)
    {
        logger.traceEntry();
        E searchedEntity = this.find(entity.getId());
        if (searchedEntity != null)
        {
            logger.traceExit(searchedEntity);
            return searchedEntity;
        }

        try
        {
            this.saveStatement(entity).executeUpdate();
        }
        catch (SQLException exception)
        {
            logger.error(exception);
            exception.printStackTrace();
        }

        E savedEntity = this.find(entity.getId());
        logger.traceExit(savedEntity);
        return savedEntity;
    }

    @Override
    public E remove(ID id) throws IllegalArgumentException
    {
        logger.traceEntry();
        if (id == null)
        {
            throw new IllegalArgumentException("id must not be null");
        }

        E searchedEntity = this.find(id);
        if (searchedEntity != null)
        {
            try
            {
                this.deleteStatement(id).executeUpdate();
            }
            catch (SQLException exception)
            {
                logger.error(exception);
                exception.printStackTrace();
            }
        }
        logger.traceExit(searchedEntity);
        return searchedEntity;
    }

    @Override
    public E update(E entity)
    {
        logger.traceEntry();
        try
        {
            this.updateStatement(entity).executeUpdate();
        }
        catch (SQLException exception)
        {
            logger.error(exception);
            exception.printStackTrace();
        }
        logger.traceExit(entity);
        return entity;
    }
}